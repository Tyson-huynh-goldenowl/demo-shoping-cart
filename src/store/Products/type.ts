export interface ProductProps {
    description: string;
    price: number;
    image: string;
    id: string;
    quality: number;
    name: string;
  }