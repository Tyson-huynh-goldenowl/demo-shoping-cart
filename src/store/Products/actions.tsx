import { PRODUCT } from "../../consts/products.const";
import {ProductProps} from './type';
export const fetchListProduct = () => {
  return {
    type: PRODUCT.fetchProductList,
  };
};

export const updateListProduct = (data: ProductProps) => {
  return {
    type: PRODUCT.updateListProduct,
    payload: data,
  };
};

export const fetchListProductSuccess = (data: any) => {
  return {
    type: PRODUCT.fetchProductListSuccess,
    payload: data,
  };
};

export const fetchListProductError = (data: any) => {
  return {
    type: PRODUCT.fetchProductListError,
    payload: data,
  };
};

export const addToCart = (data: {data:ProductProps, amount: number}) => {
  return {
    type: PRODUCT.addToCart,
    payload: data,
  };
};
