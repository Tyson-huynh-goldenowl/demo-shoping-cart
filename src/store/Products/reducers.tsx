import { PRODUCT } from "../../consts/products.const";

const initialState = {
  data: [],
  carts: [],
  amount: 0,
}

export function products(state = {
  ...initialState,
}, action: any) {
  switch (action.type) {
    case PRODUCT.fetchProductListSuccess:
      return { ...state, data: action.payload.data };
    case PRODUCT.addToCart:
      return { ...state, carts: action.payload.data, amount: action.payload.amount};
    case PRODUCT.updateListProduct:
      return {...state, data: action.payload};
    default:
      return state;
  }
}
