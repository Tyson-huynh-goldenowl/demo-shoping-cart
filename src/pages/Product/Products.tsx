import React, { useEffect } from "react";
import { Row } from "antd";
import { useDispatch, useSelector } from "react-redux";
import { fetchListProduct, addToCart, updateListProduct  } from "../../store/Products/actions";
import {ProductProps} from '../../store/Products/type';
import CardComponent from "../../components/Card";
import Menu from "../../components/Menu/Menu";

const Products = () => {
  const dispatch = useDispatch();
  const productList = useSelector((state: any) => state.products?.data);
  const carts = useSelector((state: any) => state.products?.carts);

  useEffect(() => {
    if(productList.length === 0) {
      dispatch(fetchListProduct());
    }
  }, [dispatch, productList]);

  const handleSelectItem = (id: string | undefined) => {
    let cartsSend: ProductProps[] | any = [] ;
    const product = productList.find((item: ProductProps) => item.id === id);
    if (carts.length === 0) {
      cartsSend.push({ ...product, quality: 1 });
    } else {
      const haveItem = carts.findIndex((item: ProductProps) => item.id === product.id);
      if (haveItem !== -1) {
        cartsSend = carts.map((item: ProductProps) => {
          if (item.id === product.id)
            return { ...item, quality: item.quality + 1 };
          return item;
        });

      } else {
        cartsSend = [...carts, { ...product, quality: 1 }];
      }
    }

    let totalAmount:number = 0 ;

    cartsSend.map((item: ProductProps) => {
      return totalAmount += item.price * item.quality;
    })

    dispatch(addToCart({data: cartsSend, amount:  Math.floor(totalAmount)}));

    const currentProduct = productList.map((item:ProductProps) => {
      if (item.id === product.id)
      return { ...item, quality: item.quality - 1 };
      return item;
    });

    dispatch(updateListProduct(currentProduct));
   
  };

  return (
    <>
      <Menu
        numberProduct={carts.length}
        selectedKeys="product"
      />
      <div className="site-card-border-less-wrapper">
        <Row gutter={24} justify="center">
          {productList
            ? productList.map((item: any) => {
                return (
                  <CardComponent
                    key={item.id}
                    price={item.price}
                    description={item.description}
                    imageSrc={item.image}
                    id={item.id}
                    name={item.name}
                    quality={item.quality}
                    onClickCart={handleSelectItem}
                  />
                );
              })
            : null}
        </Row>
      </div>
    </>
  );
};

export default Products;
