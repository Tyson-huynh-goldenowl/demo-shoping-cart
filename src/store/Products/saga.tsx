import { call, put, takeLatest } from "redux-saga/effects";
import { SagaIterator } from "@redux-saga/core";
import { PRODUCT } from "../../consts/products.const";
import {  fetchProducts } from "../../services/products.api";
import { fetchListProductSuccess, fetchListProductError } from "./actions";


function* getProducts() {
  yield takeLatest(PRODUCT.fetchProductList, getProductsFromAPI);
}

function* getProductsFromAPI(action: any): SagaIterator {
  try {
    // call the api
    const data = yield call(fetchProducts, {
      response: action.payload,
    });
    console.log('========>', data);
    // call the success action with data
    yield put(fetchListProductSuccess(data));
  } catch (e) {
    // call the error action with data
    yield put(fetchListProductError(e));
  }
}

export const sagas = [getProducts];
