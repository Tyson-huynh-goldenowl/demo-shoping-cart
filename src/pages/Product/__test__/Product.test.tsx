import React from "react";
import { useDispatch, useSelector } from "react-redux";
import Enzyme, { shallow } from 'enzyme';


import Products from '../Products';

describe('<Product />', () => {
    it('renders  components', () => {
      const wrapper = shallow(<Products />);
      expect(shallow).toMatchSnapshot();
    });
  
  });