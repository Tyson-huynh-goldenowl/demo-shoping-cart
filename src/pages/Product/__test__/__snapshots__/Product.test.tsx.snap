// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product /> renders  components 1`] = `[Function]`;
